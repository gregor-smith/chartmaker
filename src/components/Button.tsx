import React, { FC } from 'react'
import { cx } from 'emotion'

import { buttonStyle } from '../style'


type Props = {
    className?: string
    disabled?: boolean
    onClick?: () => void
}


const Button: FC<Props> = ({ className, ...props }) =>
    <button {...props} className={cx(buttonStyle, className)}/>


export default Button
