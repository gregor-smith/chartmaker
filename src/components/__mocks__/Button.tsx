import type { FC } from 'react'

import type { ButtonProps } from '@/components/Button'
import { cx } from 'emotion'


export const Button: FC<ButtonProps> = ({
    children,
    id,
    className,
    disabled,
    onClick
}) => {
    className = cx('mock-button', className)
    return (
        <div id={id} className={className} onClick={onClick}>
            {`Disabled: ${disabled}`}
            {children}
        </div>
    )
}
