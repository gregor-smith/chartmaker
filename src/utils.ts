import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
