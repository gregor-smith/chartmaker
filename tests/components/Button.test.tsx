import React, { createRef, MutableRefObject } from 'react'
import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import Button from '../../src/components/Button'
import { setRenderContainer, clearRenderContainer, fireEvent } from '../utils'


const container: MutableRefObject<HTMLElement | null> = createRef()
beforeEach(() => setRenderContainer(container))
afterEach(() => clearRenderContainer(container))


test('renders as button', () => {
    render(
        <Button className='test-class' disabled>
            Test children
        </Button>,
        container.current
    )

    expect(container.current).toMatchSnapshot()
})


test('click event calls onClick prop', () => {
    const mock = jest.fn<void, []>()

    render(
        <Button onClick={mock}>
            Test children
        </Button>,
        container.current
    )

    act(() => fireEvent('click', container.current?.firstChild))

    expect(mock).toHaveBeenCalledTimes(1)
})
