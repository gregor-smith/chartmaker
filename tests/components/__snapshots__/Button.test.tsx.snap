// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as button 1`] = `
.emotion-0 {
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
}

<div
  id="test-render-container"
>
  <button
    class="emotion-0 test-class"
    disabled=""
  >
    Test children
  </button>
</div>
`;
